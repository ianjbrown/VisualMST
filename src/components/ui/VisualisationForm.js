import { Form, Button } from "react-bootstrap";

function VisualisationForm(props) {

  function verticesChangeHandler(event) {
    event.preventDefault();
    props.onVerticesChange(event.target.value);
  }

  function startingChangeHandler(event) {
    event.preventDefault();
    props.onStartingChange(event.target.value);
  }

  function genImpChangeHandler(event) {
    event.preventDefault();
    props.onGenImpChange(event.target.value);
  }

  function fileChangeHandler(event) {
    event.preventDefault();
    props.onFileChange(event.target.files[0]);
  }

  function submitHandler(event) {
    event.preventDefault();
    props.onSubmit(event);
  }

  return (
    <Form data-testid="form" onSubmit={submitHandler}>
      <Form.Group className="mb-3">
        <Form.Label htmlFor="vertices">Number of Vertices</Form.Label>
        <Form.Control
          onChange={verticesChangeHandler}
          type="number"
          min="5"
          max="10"
          required
          id="vertices"
          defaultValue="5"
        />
      </Form.Group>
      {props.alg === "prim" && (
        <Form.Group className="mb-3">
          <Form.Label htmlFor="startingvertex">Select Starting Vertex</Form.Label>
          <Form.Control
            onChange={startingChangeHandler}
            type="number"
            min={0}
            max={parseInt(props.noOfVertices) - 1}
            required
            id="startingvertex"
            defaultValue="0"
          />
        </Form.Group>
      )}
      <Form.Group className="mb-3">
        <Form.Label htmlFor="select">Generate or Import Graph?</Form.Label>
        <Form.Select id="select" onChange={genImpChangeHandler}>
          <option>Select</option>
          <option>Generate Graph</option>
          <option>Import Graph</option>
        </Form.Select>
      </Form.Group>
      {props.genImp === "Import Graph" && (
        <Form.Group className="mb-3">
          <Form.Label htmlFor="file">Graph File</Form.Label>
          <Form.Control id="file" type="file" name="file" onChange={fileChangeHandler} />
        </Form.Group>
      )}
      <Form.Group className="mb-3">
        <Button className="mt-2" variant="primary" type="submit">
          Submit
        </Button>
      </Form.Group>
    </Form>
  );
}

export default VisualisationForm;
