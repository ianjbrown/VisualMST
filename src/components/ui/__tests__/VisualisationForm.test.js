import VisualisationForm from "../VisualisationForm";
import {
  render,
  screen,
  cleanup,
  waitFor,
} from "@testing-library/react";
import "@testing-library/jest-dom/";
import userEvent from "@testing-library/user-event";
import renderer from "react-test-renderer";
import five from "../../../datastructures/text/5.txt";

afterEach(() => {
  cleanup();
});

test("Visualisation Form matches snapshot", () => {
  const tree = renderer.create(<VisualisationForm />).toJSON();
  expect(tree).toMatchSnapshot();
});

test("Visualisation Form loads", () => {
  render(<VisualisationForm />);
  const form = screen.queryByTestId("form");
  expect(form).toBeInTheDocument();
});

test("Number of vertices field loads and range of input behaves as expected", () => {
  render(
    <VisualisationForm onVerticesChange={jest.fn()} onSubmit={jest.fn()} />
  );
  const form = screen.queryByTestId("form");
  const field = screen.queryByLabelText("Number of Vertices");
  const submit = screen.getByRole("button", { type: "submit" });

  expect(field).toBeInTheDocument();

  for (let i = 0; i < 20; i++) {
    userEvent.type(field, `{backspace}{backspace}${i}`);
    expect(field.value).toBe(`${i}`);
    userEvent.click(submit);
    if (i >= 5 && i <= 10) {
      expect(form).toBeValid();
    } else {
      expect(form).toBeInvalid();
    }
  }
});

test("Starting vertex field doesn't appear for Kruskal's", () => {
  render(<VisualisationForm alg="kruskal" />);
  const field = screen.queryByLabelText("Select Starting Vertex");
  expect(field).not.toBeInTheDocument();
});

test("Starting vertex field loads for Prim's and range of input behaves as expected", () => {
  render(
    <VisualisationForm
      onStartingChange={jest.fn()}
      onSubmit={jest.fn()}
      noOfVertices="6"
      alg="prim"
    />
  );
  const form = screen.queryByTestId("form");
  const field = screen.getByLabelText("Select Starting Vertex");
  const submit = screen.getByRole("button", { type: "submit" });
  expect(field).toBeInTheDocument();

  for (let i = 0; i < 20; i++) {
    userEvent.type(field, `{backspace}{backspace}${i}`);
    expect(field.value).toBe(`${i}`);
    userEvent.click(submit);
    if (i < 6) {
      expect(form).toBeValid();
    } else {
      expect(form).toBeInvalid();
    }
  }
});

test("Selection field for gen/import graph loads, file field loads when import is selected", async () => {
  const { rerender } = render(<VisualisationForm onGenImpChange={jest.fn()} />);
  const field = screen.getByLabelText("Generate or Import Graph?");
  let fileField = screen.queryByLabelText("Graph File");
  expect(field).toBeInTheDocument();
  expect(fileField).not.toBeInTheDocument();

  userEvent.selectOptions(
    field,
    screen.getByRole("option", { name: "Import Graph from file" })
  );
  rerender(
    <VisualisationForm onGenImpChange={jest.fn()} genImp="Import Graph from file" />
  );
  fileField = screen.queryByLabelText("Graph File");
  await waitFor(() => {
    expect(fileField).toBeInTheDocument();
  });
});

test("test file upload works as expected", () => {
    render(<VisualisationForm onFileChange={jest.fn()} genImp="Import Graph from file" />);
    const fileField = screen.queryByLabelText("Graph File");
    userEvent.upload(fileField, five);
    expect(fileField.files[0]).toBe(five);
})

// test("Alert displayed on invalid file input", () => { 

// })
