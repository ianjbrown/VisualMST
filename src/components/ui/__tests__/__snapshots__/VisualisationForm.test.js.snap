// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Visualisation Form matches snapshot 1`] = `
<form
  className=""
  data-testid="form"
  onSubmit={[Function]}
>
  <div
    className="mb-3"
  >
    <label
      className="form-label"
      htmlFor="vertices"
    >
      Number of Vertices
    </label>
    <input
      className="form-control"
      defaultValue="5"
      id="vertices"
      max="10"
      min="5"
      onChange={[Function]}
      required={true}
      type="number"
    />
  </div>
  <div
    className="mb-3"
  >
    <label
      className="form-label"
      htmlFor="select"
    >
      Generate or Import Graph?
    </label>
    <select
      className="form-select"
      id="select"
      onChange={[Function]}
    >
      <option>
        Select
      </option>
      <option>
        Generate Graph
      </option>
      <option>
        Import Graph
      </option>
    </select>
  </div>
  <div
    className="mb-3"
  >
    <button
      className="mt-2 btn btn-primary"
      disabled={false}
      type="submit"
    >
      Submit
    </button>
  </div>
</form>
`;
